src/app/shell/personal-cabinet/shared-cabinet/messages/chat/chat.component.ts

```diff
@@ -104,34 +151,85 @@ export class ChatComponent implements OnInit, AfterViewInit, OnDestroy {
     );
   }
 
-  addListeners() {
+  private setChatSubscriptions(): void {
     this.chatRoom$
       .pipe(
         filter((chatRoom: ChatRoom) => !!chatRoom),
         takeUntil(this.destroy$)
       )
       .subscribe((chatRoom: ChatRoom) => {
         this.chatRoom = chatRoom;
-        this.store.dispatch(
-          new GetChatRoomMessages(this.chatRoom.id, this.messagesParameters)
-        );
+        this.store.dispatch(new GetChatRoomMessages(this.chatRoom.id, this.userRole, this.messagesParameters));
+        this.getChatMembersNames();
       });
+
     this.messages$
       .pipe(
         filter((messages: Message[]) => !!messages),
+        map((messages: Message[]) => messages.reverse()),
         takeUntil(this.destroy$)
       )
       .subscribe((messages: Message[]) => {
-        this.messages = messages;
+        this.addMessages(messages);
       });
   }
 
-  onBack() {
+  private addMessages(messages: Message[]): void {
+    const isUserDown =
+      this.chatEl.nativeElement.scrollTop === this.chatEl.nativeElement.scrollHeight - this.chatEl.nativeElement.clientHeight;
+
+    if (this.isHistoryLoading) {
+      this.messages = [...messages, ...this.messages];
+      this.isHistoryLoading = false;
+    } else {
+      this.messages.push(...messages);
+    }
+
+    if (isUserDown) {
+      this.scrollDown();
+    }
+  }
+
+  onSendMessage(): void {
+    if (this.hubConnection.state === signalR.HubConnectionState.Connected && !!this.messageControl.value) {
+      const sendMessage = new SendMessage(this.chatRoom.workshopId, this.chatRoom.parentId, this.messageControl.value);
+      //TODO: Add the sender to the mailing list (backend)
+      //TODO: Remove .then() after adding sender to mailing list
+      this.hubConnection.invoke('SendMessageToOthersInGroupAsync', JSON.stringify(sendMessage)).then(() => {
+        if (this.chatRoom.id) {
+          this.store.dispatch(new GetChatRoomMessages(this.chatRoom.id, this.userRole, { from: 0, size: 1 }));
+        } else {
+          this.messages.push(
+            new Message({
+              id: '',
+              chatRoomId: '',
+              createdDateTime: Date.now().toString(),
```

Comment:
as I know, back-end generates a create date, it's not your responsibility. Only when a message is id DB, then it generates time ---> createdDateTime: Date.now().toString(),